### PR TITLE
Address possible metadata duplication in 8.1 (backport)

### PR DESCRIPTION
**Pull request by jorgeatorres**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR applies a subset of the changes from #40129, which -among other things-, prevent duplication of metadata when using the `save_meta_data()` order method, which is common with 3rd party code (including popular plugins such as WC Subscriptions).

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure both HPOS and sync are enabled in WooCommerce > Settings > Advanced > Features.
1. Run the following PHP snippet via WP-CLI (for example):
   ```php
   $order = wc_create_order();
   $order->add_meta_data( 'this_gets_duplicated', 'for reals' );
   $order->save_meta_data();
   
   echo $order->get_id();
   ```
   Take note of the ID that gets printed to the screen.
1. Make sure that metadata `this_gets_duplicated` appears only once in the output of the following WP-CLI commands:
   1. `wp db query "SELECT * FROM $(wp db prefix)wc_orders_meta WHERE order_id = ORDER_ID"`
   1. `wp post meta list ORDER_ID`
   
   Replace `ORDER_ID` with the ID from step 2.

In 8.1 (without this fix), you'll see that steps 3 shows you the metadata duplication.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php [hunk 1]
```diff
@@ -2812,12 +2821,13 @@ public function delete_meta( &$object, $meta ) {
 	 * @return int|bool  meta ID or false on failure
 	 */
 	public function add_meta( &$object, $meta ) {
-		$add_meta = $this->data_store_meta->add_meta( $object, $meta );
-		$this->after_meta_change( $object, $meta );
+		$add_meta        = $this->data_store_meta->add_meta( $object, $meta );
+		$meta->id        = $add_meta;
+		$changes_applied = $this->after_meta_change( $object, $meta );
```
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:2826]: I'm not sure it's essential, but should we do anything if this returns false? Atm I don't believe we do anything with `$changes_applied`.

Comment (barryhughes) [reply plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:2826]: Sorry, deleting ... I meant to apply this comment to line 2804.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php [hunk 2]
```diff
@@ -2791,10 +2791,19 @@ public function read_meta( &$object ) {
...
+			$db_meta = $this->data_store_meta->get_metadata_by_id( $meta->id );
+			if ( $db_meta ) {
+				$meta->key   = $db_meta->meta_key;
+				$meta->value = $db_meta->meta_value;
+			}
+		}
 
-		if ( $object instanceof WC_Abstract_Order && $this->should_backfill_post_record() ) {
+		$delete_meta     = $this->data_store_meta->delete_meta( $object, $meta );
+		$changes_applied = $this->after_meta_change( $object, $meta );
```
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:2804]: I'm not sure it's essential, but should we do anything if this returns false? Atm I don't believe we do anything with `$changes_applied`.

**Review by barryhughes — APPROVED** [review]
I think we're good, so provisionally approving (did leave one question, though).

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php (hunk 2 above)
Comment (vedanshujain) [reply plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:2804]: yeah, this should be included, adding a small commit for this.

**Comment by vedanshujain** [comment]
Seems like PHP 7.4 tests with nightly are failing because of something in the build process. I'd recommend not fixing them here as this PR targets the release branch.

**Review by peterfabian — APPROVED** [review]
LGTM, cross-checked with #40129.

**Pull request merged**